File: src/main/java/model/Snake.java
package model;

import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

import model.strategy.SnakeStrategy;

public class Snake implements Member {
    private Deque<Coordinate> bodyItems = new LinkedList<>();
    private SnakeStrategy strategy;
    private ModifiableArena arena;
    private String name;

    public Snake(ModifiableArena arena, SnakeStrategy strategy, String name) {
        this.arena = arena;
        this.strategy = strategy;
        this.name = name;
        this.bodyItems.add(arena.generateRandomFreeCoordinate());
    }

    public List<Coordinate> getBodyItems() {
        return new ArrayList<>(bodyItems);
    }

    public Coordinate getHeadCoordinate() {
        return bodyItems.getFirst();
    }

    protected Coordinate decideNextCoordinate() {
        Direction direction = strategy.nextMove(this, arena);
        Coordinate nextCoordinate = arena.nextCoordinate(bodyItems.getFirst(), direction);
        return nextCoordinate;
    }

    protected void moveTo(Coordinate nextCoordinate) {
        if (!arena.isFood(nextCoordinate)) {
            bodyItems.removeLast();
        } else {
            arena.removeFood(nextCoordinate);
        }
        if (arena.isOccupied(nextCoordinate)) {
            System.out.println("Snake died: " + name);
            throw new SnakeDeadException();
        }
        bodyItems.addFirst(nextCoordinate);
    }

    public Coordinate getTailCoordinate() {
        return bodyItems.getLast();
    }

    @Override
    public boolean occupies(Coordinate nextCoordinate) {
        return bodyItems.contains(nextCoordinate);
    }

    public int length() {
        return bodyItems.size();
    }

    public String getName() {
        return name;
    }

    @Override
    public String toString() {
        return "Snake [bodyItems=" + bodyItems + ", strategy=" + strategy + "]";
    }
}

File: snake-framework/src/main/kotlin/co/electric/snake/framework/model/Snake.kt
package co.electric.snake.framework.model

import co.electric.snake.framework.strategy.SnakeStrategy
import org.slf4j.LoggerFactory
import java.util.*

open class Snake(private val arena: ModifiableArena, private val strategy: SnakeStrategy, val name: String) : Member {

    companion object {
        private val LOG = LoggerFactory.getLogger(Snake::class.java)

        private const val SNAKE_DIED_LOG_MESSAGE = "Snake died: {}"
    }

    private val bodyItems: Deque<Coordinate> = LinkedList()

    init {
        bodyItems.add(arena.generateRandomFreeCoordinate())
    }

    override fun occupies(coordinate: Coordinate): Boolean {
        return bodyItems.contains(coordinate)
    }

    protected fun decideNextCoordinate(): Coordinate {
        val direction = strategy.nextMove(this, arena)
        return arena.nextCoordinate(bodyItems.first, direction)
    }

    protected fun moveTo(coordinate: Coordinate) {
        if (!arena.isFood(coordinate)) {
            bodyItems.removeLast()
        } else {
            arena.removeFood(coordinate)
        }
        if (arena.isOccupied(coordinate)) {
            LOG.info(SNAKE_DIED_LOG_MESSAGE, name)
            throw SnakeDeadException()
        }
        bodyItems.addFirst(coordinate)
    }

    fun getBodyItemsInNewList(): List<Coordinate> {
        return ArrayList(bodyItems)
    }

    fun getHeadCoordinate(): Coordinate {
        return bodyItems.first
    }

    fun getTailCoordinate(): Coordinate {
        return bodyItems.last
    }

    fun getLength(): Int {
        return bodyItems.size
    }

    override fun toString(): String {
        return "Snake [bodyItems=$bodyItems, strategy=$strategy]"
    }

}
